akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (C) 2015-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl.fusing
+
+import akka.annotation.InternalApi
+import akka.stream.scaladsl.{ Flow, Keep, Source }
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler }
+import akka.stream._
+import akka.util.OptionVal
+
+import scala.collection.immutable
+import scala.concurrent.{ Future, Promise }
+import scala.util.control.NonFatal
+
+@InternalApi private[akka] final class FlatMapPrefix[In, Out, M](n: Int, f: immutable.Seq[In] => Flow[In, Out, M])
+    extends GraphStageWithMaterializedValue[FlowShape[In, Out], Future[M]] {
+
+  require(n >= 0, s"FlatMapPrefix: n ($n) must be non-negative.")
+
+  val in = Inlet[In](s"${this}.in")
+  val out = Outlet[Out](s"${this}.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val matPromise = Promise[M]
+    val logic = new GraphStageLogic(shape) with InHandler with OutHandler {
+      var accumulated = List.empty[In]
+
+      private var subSource = OptionVal.none[SubSourceOutlet[In]]
+      private var subSink = OptionVal.none[SubSinkInlet[Out]]
+
+      setHandlers(in, out, this)
+
+      override def postStop(): Unit = {
+        matPromise.tryFailure(new AbruptStageTerminationException(this))
+        if (subSource.isDefined && !subSource.get.isClosed)
+          subSource.get.complete()
+        if (subSink.isDefined && !subSink.get.isClosed)
+          subSink.get.cancel()
+        super.postStop()
+      }
+
+      override def onPush(): Unit = {
+        if (subSource.isDefined) {
+          subSource.get.push(grab(in))
+        } else {
+          accumulated ::= grab(in)
+          if (accumulated.size == n) {
```

Comment:
`List.size` is O(n) => use VectorBuilder :)

Reply:
notice `accumulated' is now a `scala.collection.mutable.Buffer'